core/src/main/java/hudson/PluginManager.java

```diff
@@ -1150,9 +1152,23 @@ public HttpResponse doInstallPlugins(StaplerRequest req) throws IOException {
                     index = n.indexOf('.', index + 1);
                 }
             }
+            
             if (p == null) {
                 throw new Failure("No such plugin: " + n);
             }
+            
+            // JENKINS-33308 - automatically install implied/previously bundled dependencies for older plugins that may need them
+            for(PluginWrapper.Dependency previouslyBundledDependency : ClassicPluginStrategy.getPreviouslyBundledDependencies(p.requiredCore)) {
```

Comment:
This smells like the wrong place to be doing this. The detached plugin mapping should be logically applied to incoming metadata, whether from UC JSON or from a plugin manifest. You seem to be adding in these dependencies at the last minute, whereas they should been considered dependencies from the start.

I suspect that the difference is tangible: if you go into the setup wizard and check the box next to some ancient plugin, the box for the detached dependency like `matrix-project` should get checked then too (which of course _also_ means that we will wind up installing it, but without any special handling _here_).